### ISSUE TITLE
Killing ServiceHub.Host.Node.x86.exe crashes Visual Studio RC.3

### ISSUE THREAD

**Issue by JoshVarty** [dotnet/roslyn#16859]
**Version Used**: 
![image](https://cloud.githubusercontent.com/assets/1249087/22486977/6f639d9a-e7c0-11e6-8ee8-14f7f5303ac8.png)

**Steps to Reproduce**:

1. Open Visual Studio
2. Kill `ServiceHub.Host.Node.x86.exe` from Task Manager
3. Visual Studio Crashes

Dump available at:
`\\vspfs3\Public\jovarty\dumps\devenv_crash_on_ServiceHub_killed.dmp`

Stack Trace:

```
>	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.ErrorReporting.FatalError.Report(System.Exception exception, System.Action<System.Exception> handler)	Unknown
 	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Remote.RemoteHostClientServiceFactory.RemoteHostClientService.OnConnectionChanged(object sender, bool connected)	Unknown
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Remote.RemoteHostClient.OnConnectionChanged(bool connected)	Unknown
 	Microsoft.VisualStudio.LanguageServices.Next.dll!Microsoft.VisualStudio.LanguageServices.Remote.ServiceHubRemoteHostClient.OnRpcDisconnected(object sender, StreamJsonRpc.JsonRpcDisconnectedEventArgs e)	Unknown
 	StreamJsonRpc.dll!StreamJsonRpc.JsonRpc.OnJsonRpcDisconnected(StreamJsonRpc.JsonRpcDisconnectedEventArgs eventArgs)	Unknown
 	StreamJsonRpc.dll!StreamJsonRpc.JsonRpc.ReadAndHandleRequestsAsync()	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.InvokeMoveNext(object stateMachine)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.Run()	Unknown
 	mscorlib.dll!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action action, bool allowInlining, ref System.Threading.Tasks.Task currentTask)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.FinishContinuations()	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.FinishStageThree()	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task<System.__Canon>.TrySetResult(System.__Canon result)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder<string>.SetResult(string result)	Unknown
 	StreamJsonRpc.dll!StreamJsonRpc.DelimitedMessageHandler.ReadAsync(System.Threading.CancellationToken cancellationToken)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.InvokeMoveNext(object stateMachine)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.Run()	Unknown
 	mscorlib.dll!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action action, bool allowInlining, ref System.Threading.Tasks.Task currentTask)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.FinishContinuations()	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.FinishStageThree()	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task<System.__Canon>.TrySetResult(System.__Canon result)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder<string>.SetResult(string result)	Unknown
 	StreamJsonRpc.dll!StreamJsonRpc.HeaderDelimitedMessageHandler.ReadCoreAsync(System.Threading.CancellationToken cancellationToken)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.InvokeMoveNext(object stateMachine)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.Run()	Unknown
 	mscorlib.dll!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action action, bool allowInlining, ref System.Threading.Tasks.Task currentTask)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.FinishContinuations()	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.FinishStageThree()	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task<System.Threading.Tasks.VoidTaskResult>.TrySetResult(System.Threading.Tasks.VoidTaskResult result)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder<System.Threading.Tasks.VoidTaskResult>.SetResult(System.Threading.Tasks.VoidTaskResult result)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder<System.Threading.Tasks.VoidTaskResult>.SetResult(System.Threading.Tasks.Task<System.Threading.Tasks.VoidTaskResult> completedTask)	Unknown
 	StreamJsonRpc.dll!StreamJsonRpc.ReadBufferingStream.FillBufferAsync(System.Threading.CancellationToken cancellationToken)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.InvokeMoveNext(object stateMachine)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.Run()	Unknown
 	mscorlib.dll!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action action, bool allowInlining, ref System.Threading.Tasks.Task currentTask)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.FinishContinuations()	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.FinishStageThree()	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task<int>.TrySetResult(int result)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder<int>.SetResult(int result)	Unknown
 	Microsoft.ServiceHub.Client.dll!Microsoft.ServiceHub.Utility.WrappedStream.ReadAsync(byte[] buffer, int offset, int count, System.Threading.CancellationToken cancellationToken)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.InvokeMoveNext(object stateMachine)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.Run()	Unknown
 	mscorlib.dll!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action action, bool allowInlining, ref System.Threading.Tasks.Task currentTask)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.FinishContinuations()	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.FinishStageThree()	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task<int>.TrySetResult(int result)	Unknown
 	mscorlib.dll!System.Threading.Tasks.TaskFactory<int>.FromAsyncTrimPromise<System.IO.Stream>.Complete(System.IO.Stream thisRef, System.Func<System.IO.Stream, System.IAsyncResult, int> endMethod, System.IAsyncResult asyncResult, bool requiresSynchronization)	Unknown
 	mscorlib.dll!System.Threading.Tasks.TaskFactory<int>.FromAsyncTrimPromise<System.IO.Stream>.CompleteFromAsyncResult(System.IAsyncResult asyncResult)	Unknown
 	System.Core.dll!System.IO.Pipes.PipeStream.AsyncPSCallback(uint errorCode, uint numBytes, System.Threading.NativeOverlapped* pOverlapped)	Unknown
 	mscorlib.dll!System.Threading._IOCompletionCallback.PerformIOCompletionCallback(uint errorCode, uint numBytes, System.Threading.NativeOverlapped* pOVERLAP)	Unknown
```

**Comment by JoshVarty** [comment]
@heejaechang Pilchie mentioned that I should bring this up with you. Was this the result of an intentional design decision? Is it possible to handle crashed ServiceHub instances without bringing down Visual Studio?

**Comment by heejaechang** [comment]
@JoshVarty Yes it was. but since we made the decision, we have been having some customer feedbacks on the crashing so, tagging some people to see whether we still want to keep this decision or want to tweak it a bit.

tagging other people @CyrusNajmabadi @MattGertz @srivatsn @Pilchie @jasonmalinowski

**Comment by heejaechang** [comment]
@JoshVarty if you kill the process, then you probably make VS 1. lose some features, 2. make it unstable.

if we are not crashing VS, what would be your expectation?

1. VS starts run features that used to run in OOP (service hub) in proc? (which can tank VS performance or crash due to out of memory)
2. VS shows some warning to users and lose some features but keep running. (or provide a way to start the OOP)
3. VS just recover if possible and restart OOP by itself.
4. Keep current behavior.

**Comment by MattGertz** [comment]
Any changes to this are very unlikely to meet the bar for RTW now.

**Comment by JoshVarty** [comment]
Ideally `3.` would be the desired behavior.

Historically (Dev14) ServiceHub sometimes locked files for me and doesn't always shut down when I close Visual Studio. So occasionally when I see VS complain about locked files I'll usually start trouble shooting by killing ServiceHub.

**Comment by heejaechang** [comment]
@JoshVarty I see. I wonder how other think.

@MattGertz we could change behavior in update 1.

just FYI. I think some people might be against (3). because 3 is actually quite hard to do and might left VS unstable. when process is forcefully terminated, it might be quite hard for us to guard from it in all places and recover from it especially if some operation was on-going.

but probably most of case, it would be just fine.

**Comment by JoshVarty** [comment]
What features are currently executed out of process? I know Find All References is one of them.

I'm not familiar with the OOP implementation but my expectation would be that any ongoing actions get cancelled and state is rebuilt from scratch. Is there any OOP state that **must** be recreated in the event of a crash?

**Comment by heejaechang** [comment]
@JoshVarty what it currently do is not as matter as what requirement it will impose on all features that might be implemented in OOP.

in current model, features only need to care about cancellation. 

once we let OOP to be terminated and vs left alive, VS side code must know how to deal with 1. not crash on such case, 2. handle on OOP not available case.

once we let OOP to recover, features needs to 1. either become stateless features 2. know how to recover states for features to work.

even after that, VS need to figure out what we do on the on-going operations that got aborted due to the OOP termination. show info bar? debug output? what happen if it was something like file change, undo file change?

basically, feature author that want to run things in OOP gets way more area to cover to make feature to run on OOP.

in current model, they only need to care only 1 thing. cancellation.

**Comment by davkean** [comment]
Coincidentally, I'm seeing this crash in the wild: https://github.com/dotnet/roslyn/issues/16859, for example: https://github.com/dotnet/roslyn/issues/16867

**Comment by heejaechang** [comment]
@davkean yep. I have seen that as well. that is why I wonder whether we want to change our decision on the behavior or not.

**Comment by MattGertz** [comment]
@heejaechang @srivatsn @panopticoncentral @tmeschter I've asked HeeJae & Paul to refocus on the OOP crashes in general.  **For this issue specifically**, given that we are seeing a rapidly rising count of hits, I'd like an RTW solution that does not crash VS, but instead provides some record that the service was shut down (ie. non-fatal Watson).

**Comment by heejaechang** [comment]
@MattGertz if we don't crash the VS, what should we do for VS? show users that VS might be unstable? some features such as code lens and etc won't work anymore?

**Comment by MattGertz** [comment]
Can we do InfoBar for now?  We can work out something better in subsequent releases.

**Comment by heejaechang** [comment]
@MattGertz we can, but now question is vs will be in unstable mode.

**Comment by MattGertz** [comment]
@heejaechang Yes, I understand that.  That's what the InfoBar message will say -- something like "Something bad happened, we recommend saving your work & restarting."  That way at least people won't lose stuff, and we'll still have the NFW to investigate.  Work with PM on the exact message.  Loc freeze is Friday, so need this before then.

**Comment by srivatsn** [comment]
Agreed. I think the message should say something about the process dying so that they can associate the killing of the process with the message.

**Comment by heejaechang** [comment]
alright then.

**Comment by heejaechang** [comment]
@morrisjoe are you still our PM?

**Comment by heejaechang** [comment]
@MattGertz @srivatsn @morrisjoe another question. once OOP is gone, and we show users info bar.

if user decided to keep us VS, what should we do for features that used to run in OOP.

1. restart OOP and let users to keep use the features.
2. let users to use the features in proc.
3. disable all those features.

**Comment by MattGertz** [comment]
@heejaechang What do you recommend, based on cost & risk?  You've listed them in my preferred order.

**Comment by MattGertz** [comment]
BTW, the string for the InfoBar can be brought to Shiproom before the actual OOP change -- that way, it will give us a buffer on the localization work.  Work with @morrisjoe on that.

**Comment by heejaechang** [comment]
@MattGertz 

First, whatever I do, there will be a chance VS might still crash when OOP goes away. not because I explicitly kill VS, but because some unpredictable unhandled exception caused by forcefully closed connection to OOP.

(2) is the cheapest. problem is it can cause OOM (FSA, codelens, LUT all running in proc)

(1) and (3) are about same work. 

but (3) will be safer since it doesn't try to continue features in unknown state.

(1) will work for some features, won't work for others depends on how features are written.

**Comment by MattGertz** [comment]
I will ask Shiproom for guidance in a few minutes and will loop back.

**Comment by srivatsn** [comment]
I don't think we should do (2). That opens up a support vector - if features continue to work even if I kill that other process, then we have always to test both scenarios and ensure both works. I think it's safest to do (3) or option (4) show info bar and do nothing. VS will be unstable and it's best for the user to restart anyway.

**Comment by heejaechang** [comment]
@srivatsn if we do nothing, then (2) happens.

**Comment by MattGertz** [comment]
@heejaechang Second option.

**Comment by MattGertz** [comment]
Need to verify that all features work in-proc ,  though.   And please front-end string to Loc.

**Comment by heejaechang** [comment]
@MattGertz @srivatsn so decided to this?

(2) is the cheapest. problem is it can cause OOM (FSA, codelens, LUT all running in proc)

**Comment by MattGertz** [comment]
Also add a telemetry point in addition to the NFW.   Although NFW  will give us data,  a telemetry point will get to us faster & will be easier to use to watch trends.

**Comment by MattGertz** [comment]
Yes,  please go ahead.  We understand that it may OOM.

**Comment by heejaechang** [comment]
@srivatsn @MattGertz so (2) seem safer than (3) since most features doesn't know what to do when remote host return null ..

**Comment by heejaechang** [comment]
@morrisjoe just to make sure. we not only need string to show users, but a web page to explain things more in detail. just like what we did for low VM.

**Comment by MattGertz** [comment]
@heejaechang Please prepare both (2) and (3).  We'll give Shiproom a choice (and we'll need to test (3) a lot).

**Comment by heejaechang** [comment]
PR - https://github.com/dotnet/roslyn/pull/17174 - not done yet

**Comment by heejaechang** [comment]
tagging @KevinH-MS @ManishJayaswal @rynowak these people should have been brought into the discussion.

**Comment by heejaechang** [comment]
fix is now in.
